### ISSUE TITLE
It is possible to construct arrays with negative first dimension

### ISSUE THREAD

**Issue by LilithHafner** [JuliaLang/julia#54325]
Normally, invalid array sizes are caught by `Core.checked_dims`, but `Core.checked_dims(-10, 10, 10)` does not throw, leading to this error
```julia
julia> Array{Int}(undef, -10, 10, 10)
ERROR: ArgumentError: invalid GenericMemory size: too large for system address width
Stacktrace:
 [1] GenericMemory
   @ ./boot.jl:514 [inlined]
 [2] new_as_memoryref
   @ ./boot.jl:536 [inlined]
 [3] Array
   @ ./boot.jl:583 [inlined]
 [4] (Array{Int64})(::UndefInitializer, m::Int64, n::Int64, o::Int64)
   @ Core ./boot.jl:596
 [5] top-level scope
   @ REPL[64]:1
```
When "ERROR: ArgumentError: invalid Array dimensions" is more common. We should, at the very least, put an inline comment in the 3+ arg `Core.checked_dims` method that notes this gap.

**Comment by LilithHafner** [comment]
Oh, it's worse:

```julia
julia> x = Array{Int}(undef, -10, 10, 0)
-10×10×0 Array{Int64, 3}
```

**Comment by LilithHafner** [comment]
This is broken on beta1 but works (throws) on master. Probably just need to find and backport the right PR.

**Comment by giordano** [comment]
Kinda duplicate of #54244, fixed by #54255

**Comment by LilithHafner** [comment]
Exact duplicate of https://github.com/JuliaLang/julia/pull/54255#issuecomment-2078093893
